packages/commonwealth/server/controllers/server_groups_methods/refresh_community_memberships.ts

```diff
@@ -33,28 +38,133 @@ export async function __refreshCommunityMemberships(
       },
     },
     attributes: ['id', 'address'],
+    include: {
+      model: this.models.Membership,
+      as: 'Memberships',
+      required: false,
+    },
   });
 
-  const result = await Bluebird.map(
-    addresses,
-    (address) => {
-      return refreshMembershipsForAddress(
-        this.models,
-        this.tokenBalanceCache,
+  const toCreate = [];
+  const toUpdate = [];
+
+  const processMembership = async (
+    address: AddressAttributes,
+    currentGroup: GroupAttributes,
+  ) => {
+    const existingMembership = address.Memberships.find(
+      ({ group_id }) => group_id === currentGroup.id,
+    );
+    if (existingMembership) {
+      // membership exists
+      const expiresAt = moment(existingMembership.last_checked).add(
+        MEMBERSHIP_TTL_SECONDS,
+        'seconds',
+      );
+      if (moment().isBefore(expiresAt)) {
+        // membership is fresh, do nothing
+        return;
+      }
+      // membership stale, update
+      const computedMembership = await refreshAndQueueOperation(
         address,
-        groups,
+        currentGroup,
+      );
+      toUpdate.push(computedMembership);
+      return;
+    }
+
+    // membership does not exist, create
+    const computedMembership = await refreshAndQueueOperation(
+      address,
+      currentGroup,
+    );
+    toCreate.push(computedMembership);
+  };
+
+  const refreshAndQueueOperation = async (
+    address: AddressAttributes,
+    currentGroup: GroupAttributes,
+  ) => {
+    const { requirements } = currentGroup;
+    const { isValid, messages } = await validateGroupMembership(
+      address.address,
+      requirements,
+      this.tokenBalanceCache,
+    );
+    const computedMembership = {
+      group_id: currentGroup.id,
+      address_id: address.id,
+      reject_reason: isValid ? null : JSON.stringify(messages),
+      last_checked: Sequelize.literal('CURRENT_TIMESTAMP') as any,
+    };
+    return computedMembership;
+  };
+
+  console.log(
+    `Checking ${addresses.length} addresses in ${groupsToUpdate.length} groups in ${community.id}...`,
+  );
+
+  await Bluebird.map(
+    groupsToUpdate,
+    async (currentGroup) => {
+      return Bluebird.map(
+        addresses,
+        async (address) => {
+          return processMembership(address, currentGroup);
+        },
+        {
+          concurrency: 20,
+        },
       );
     },
-    { concurrency: 300 },
+    { concurrency: 20 },
   );
 
-  const allMemberships = flatten(result);
+  console.log(`Done checking. Starting create...`);
+
+  // first create new rows
+  await this.models.Membership.bulkCreate(toCreate);
+
+  console.log(`Starting update...`);
+
+  // TODO: test this query out!!!
```

Comment:
tested?

Reply:
May not need testing -- likely never gets triggered.